### ISSUE TITLE
Guidance on Training from Scratch or Fine-Tuning

### ISSUE THREAD

**Issue by alfausa1** [ZhengPeng7/BiRefNet#164]
Hi,
I would like to ask how many images you would recommend for training a model from scratch, and what weights you would suggest starting with.

My use case is object segmentation on plain backgrounds. The general model currently works quite well for most cases, but there are a few specific scenarios that could be improved. This is why I’m considering training or fine-tuning.

I have a dataset of around 7,000 images at 2K resolution. What would you recommend in this case?

Thank you in advance for your help!

**Comment by ZhengPeng7** [comment]
For common cases with no extremely complicated shapes, 500-1,000 images should be enough for training from scratch.
If your cases are very different from the training sets I used to train the general version weights, I suggest training from scratch when you have enough images. Otherwise, fine-tuning could be a better way.

In your case, I recommend training from scratch. BTW, you can check the model efficiency part in README; use FP16 + compile==True + PyTorch==2.5.1 to try to save GPU memory to do less downscaling on your 2K data.

**Comment by Roshan-digi5** [comment]
Hello,

First of all, thank you for your incredible work and contributions!

I want to train a model specifically for removing backgrounds from car images. I have a dataset of approximately 80,000 images. Could you guide me on the best practices to follow, which model and settings would be most suitable, and whether there are any tutorials available for training or fine-tuning a model?

**Comment by ZhengPeng7** [comment]
I've made a [guideline](https://github.com/ZhengPeng7/BiRefNet?tab=readme-ov-file#pen-fine-tuning-on-custom-data) of fine-tuning in my README. For settings of fine-tuning, you can use the default settings except for the `epochs`. If you still have a problem after following it, plz tell me.

**Comment by Roshan-digi5** [comment]
Thank you will let you know in case of any issue.

**Comment by alfausa1** [comment]
Hi,

Thank you so much for taking the time to reply! 

I wanted to ask specifically about the configurations, losses and backbone you would recommend for my use case. Are there any particular hyperparameters or architectures you find especially suitable for this type of task? Any additional guidance would be greatly appreciated.

Thanks again for your support!

**Comment by alfausa1** [comment]
Sorry for not specifying earlier, my use case is object segmentation on a plain background (not cars). Many objects do have transparencies and some small details like tiny holes.

**Comment by ZhengPeng7** [comment]
That would be a **general** case. I'm not sure about it (otherwise, I would have added the updates to the default settings).

**Comment by ZhengPeng7** [comment]
If you find it works worse on some specific cases, training only on them would help a lot. Hard negative samples usually teach the model more about it.

Yeah, following the setting there with `compile`, `FP16`, and `batch_size == 2`, the training would take ~25GB.

**Comment by Roshan-digi5** [comment]
I'm following the guidelines you created but still unable to understand i have updated my dataset paths in same way as there said till step2 after that what changes needs to be done in config.py as well as in train.py any more guidance or any Collab demo for fine-tunning

**Comment by ZhengPeng7** [comment]
OK, thanks for the suggestion. I'll try to record a video of ~1 min to start basic fine-tuning.

**Comment by Roshan-digi5** [comment]
Thank you!

**Comment by Roshan-digi5** [comment]
@alfausa1 is it possible for you to share your model or data i want to try your case as well

**Comment by ZhengPeng7** [comment]
Oh, by the way, now, with the latest codes, the standard BiRefNet can be trained on a single 4090; you can check the FP16 GPU mem part in README.

**Comment by Roshan-digi5** [comment]
okay thank you

**Comment by Roshan-digi5** [comment]
do you think you will be doing video or a colab for training soon ?

**Comment by Roshan-digi5** [comment]
I've created a large dataset by combining data provided by you with my own collection. Now, I want to train a model that performs well not only on plain backgrounds but also on more detailed items such as jewelry and clothing. What would you recommend as the best approach to achieve strong performance across both these scenarios?

**Comment by ZhengPeng7** [comment]
If it's a 0 or 1 segmentation task, use the correct loss in the setting. Increasing the IoU fine-tuning epochs may do some improvement on it.
The most important thing is to make the training data as close as possible to your practical cases.

**Comment by ZhengPeng7** [comment]
Hey guys, I've uploaded the tutorial on fine-tuning BiRefNet with custom data to my YouTube channel: https://youtu.be/FwGT_0V9E-k

Let me know if you still have questions after watching the screen recording.

**Comment by Jaga-2410** [comment]
Hi @ZhengPeng7, I have a dataset of 9600+ train and 2400+ validation images. Both of these includes original and ground truth. I recently watched your video on youtube and tried to train the BiRefNet-general-epoch_244.pth with my dataset I mentioned earlier. I did all the configurations you had done in the video and finally when I tried to run this command:

(birefnetenv) ubuntu@ip-172-0-21-136:~/Birefnet-Train$ ./train_test.sh ft-recording 0,1 0
-bash: ./train_test.sh: /bin/sh^M: bad interpreter: No such file or directory

I ran into a error as you can see. Can you point out what was wrong or is there any other way to start the training with train.py instead of train_test.sh. Thanks in advance!

**Comment by ZhengPeng7** [comment]
It's easy to solve. You can change the `/bin/sh` in the first line to `/bin/bash` and everything should be okay.

**Comment by Jaga-2410** [comment]
Thank you @ZhengPeng7. I'll try it and get back to you.
